pow/src/main/java/tech/pegasys/artemis/pow/DepositProcessingController.java

```diff
@@ -0,0 +1,210 @@
+  private volatile boolean active = false;
+
+  // BlockByBlock mode is used to request deposit events and block information for each block
+  private volatile boolean isBlockByBlockModeOn = false;
+
+  private BigInteger latestSuccessfullyQueriedBlock = BigInteger.ZERO;
+  private BigInteger latestCanonicalBlockNumber = BigInteger.ZERO;
+
+  public DepositProcessingController(
+      Eth1Provider eth1Provider,
+      Eth1EventsChannel eth1EventsChannel,
+      AsyncRunner asyncRunner,
+      DepositFetcher depositFetcher) {
+    this.eth1Provider = eth1Provider;
+    this.eth1EventsChannel = eth1EventsChannel;
+    this.asyncRunner = asyncRunner;
+    this.depositFetcher = depositFetcher;
+  }
+
+  public synchronized void switchToBlockByBlockMode() {
+    LOG.debug("Switching to blockByBlock mode");
+    isBlockByBlockModeOn = true;
+  }
+
+  // inclusive of start block
+  public synchronized void startSubscription(BigInteger subscriptionStartBlock) {
+    LOG.debug("Starting subscription at block {}", subscriptionStartBlock);
+    latestSuccessfullyQueriedBlock = subscriptionStartBlock.subtract(BigInteger.ONE);
+    newBlockSubscription =
+        eth1Provider
+            .getLatestBlockFlowable()
+            .map(EthBlock.Block::getNumber)
+            .map(number -> number.subtract(ETH1_FOLLOW_DISTANCE.bigIntegerValue()))
+            .subscribe(this::onNewCanonicalBlockNumber, this::onSubscriptionFailed);
+  }
+
+  public void stopIfSubscribed() {
+    if (newBlockSubscription != null) {
+      newBlockSubscription.dispose();
+    }
+  }
+
+  // Inclusive
+  public synchronized SafeFuture<Void> fetchDepositsFromGenesisTo(BigInteger toBlockNumber) {
+    return depositFetcher.fetchDepositsInRange(BigInteger.ZERO, toBlockNumber);
+  }
+
+  private synchronized void onSubscriptionFailed(Throwable err) {
+    Disposable subscription = newBlockSubscription;
+    if (subscription != null) {
+      subscription.dispose();
+    }
+    LOG.warn("New block subscription failed, retrying.", err);
+    asyncRunner
+        .getDelayedFuture(Constants.ETH1_SUBSCRIPTION_RETRY_TIMEOUT, TimeUnit.SECONDS)
+        .finish(
+            () -> startSubscription(latestSuccessfullyQueriedBlock),
+            (error) ->
+                LOG.warn(
+                    "Unable to subscribe to the Eth1Node. Node won't "
+                        + "have access to new deposits after genesis.",
+                    error));
+  }
+
+  private synchronized void onNewCanonicalBlockNumber(BigInteger latestCanonicalBlockNumber) {
+    if (latestCanonicalBlockNumber.compareTo(this.latestCanonicalBlockNumber) <= 0) {
+      return;
+    }
+    this.latestCanonicalBlockNumber = latestCanonicalBlockNumber;
+    fetchLatestSubscriptionDeposits();
+  }
+
+  private synchronized void fetchLatestSubscriptionDeposits() {
+    if (isBlockByBlockModeOn) {
+      fetchLatestDepositsOneBlockAtATime();
+    } else fetchLatestSubscriptionDepositsOverRange();
```

Comment:
We should consistently use braces.
```suggestion
    } else {
      fetchLatestSubscriptionDepositsOverRange();
    }
```

Reply:
Done.